griptape/utils/chat.py

```diff
@@ -15,23 +16,37 @@ class Chat:
     intro_text: Optional[str] = field(default=None, kw_only=True)
     prompt_prefix: str = field(default="Q: ", kw_only=True)
     response_prefix: str = field(default="A: ", kw_only=True)
+    streaming_output_fn: Callable[[str], None] = field(
+        default=lambda x: print(x, end=""), kw_only=True
+    )
     output_fn: Callable[[str], None] = field(
         default=Factory(lambda: print), kw_only=True
     )
 
-    def start(self) -> None:
+    def start(self, initial_turn: str = None) -> None:
+        if initial_turn:
+            self.output_fn(initial_turn)
+
         if self.intro_text:
             self.output_fn(self.intro_text)
         while True:
             question = input(self.prompt_prefix)
 
             if question.lower() in self.exit_keywords:
                 self.output_fn(self.exiting_text)
-
                 break
             else:
                 self.output_fn(self.processing_text)
 
-            self.output_fn(
-                f"{self.response_prefix}{self.structure.run(question).output.to_text()}"
-            )
+            if self.structure.stream:
+                first_chunk: bool = True
```

Comment:
@mattma1970 can you please explain this to me again? Why can't we just place `self.streaming_output_fn(f"{self.response_prefix}")` before the loop?

Reply:
It's a user experience issue. If you place it before the loop then the user will see the 'A:' prefix in the console and then have to wait until the first chunk is returned. That time to first token [chunk] can be uncomfortably long. It left me wondering if something had hung. If we include it in the loop then user will see 'processing..' which lets them believe something is happening and then the 'A: [first chunk']' appear together. 

Happy to change it if you have a different view about the UX.

Reply:
Got it, thank you for explaining! I wonder if we can use [this](https://rich.readthedocs.io/en/stable/reference/spinner.html) to display a spinner until the first chunk? We use `rich` in other places of the framework but we underutilize it.

Reply:
What do you have in mind there? As an alternative to the 'processing..' text or render the 'processing.. \n A:' and put a spinner after the A: ? 
@shhlife has a video in the Trade School where he shows that you can use rich for better command line experience. It feels like its personal taste to make use of it and what spinner to use. What if we had rich wired up in the chat utility but by default had it turned off. Users could then just pass a spinner name if they want it?

Reply:
Ok after thinking about it some more I don't think we need to overcomplicate things with `rich` right now. I was just trying to get away from the `first_chunk` pattern. Let's do this instead:

```python
stream = Stream(self.structure).run(question)
next(stream)
self.streaming_output_fn(self.response_prefix)
for chunk in stream:
    self.streaming_output_fn(chunk.value)
self.streaming_output_fn("\n")
```

Reply:
Hi. This will drop the first chunk unless we do
`self.streaming_output_fn(self.response_prefix+next(chunk.value)`

Does this still address your concerns about a 'first-token' pattern?

Reply:
Ah good catch. Let's do:

```python
stream = Stream(self.structure).run(question)
first_chunk = next(stream)
self.streaming_output_fn(self.response_prefix + first_chunk)
for chunk in stream:
    self.streaming_output_fn(chunk.value)
self.streaming_output_fn("\n")
```